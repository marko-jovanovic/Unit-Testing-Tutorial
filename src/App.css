.App {
  text-align: center;
}

.loading-spinner-wrapper {
  display: flex;
  flex-flow: column;
  align-items: center;
  justify-content: center;
}

.loading-spinner {
  width: 50px;
  height: 50px;
  background-color: white;
  border: 5px solid #bdbdbd;
  border-top: 5px solid #ee2055;
  border-radius: 50%;
  animation: loading-spinner 1.2s linear infinite;
}

@keyframes loading-spinner {
  0% {
    transform: rotateZ(0deg);
  }
  100% {
    transform: rotateZ(360deg);
  }
}
