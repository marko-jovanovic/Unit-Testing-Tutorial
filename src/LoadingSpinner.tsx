import React from "react";

const LoadingSpinner: React.FC = () => {
  return (
    <div className="loading-spinner-wrapper">
      <div className="loading-spinner"></div>
      <p className="loading-message">Please wait...</p>
    </div>
  );
};

export default LoadingSpinner;
