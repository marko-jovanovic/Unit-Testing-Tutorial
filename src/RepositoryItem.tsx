import React from "react";

interface RepositoryItemProps {
  name: string;
  watchCount: number;
  homepageUrl: string;
  repositoryUrl: string;
}

const RepositoryItem: React.FC<RepositoryItemProps> = ({
  name,
  watchCount,
  homepageUrl,
  repositoryUrl
}) => {
  return (
    <div className="repository-item">
      <p className="repo-name">
        <a href={repositoryUrl} target="_blank" rel="noopener noreferrer">
          <span>{name}</span>
          <span> ({watchCount} watchers)</span>
        </a>

        {homepageUrl && repositoryUrl && homepageUrl !== repositoryUrl && (
          <>
            <span> - </span>
            <a href={homepageUrl} target="_blank" rel="noopener noreferrer">
              Home Page
            </a>
          </>
        )}
      </p>
    </div>
  );
};

export default RepositoryItem;
