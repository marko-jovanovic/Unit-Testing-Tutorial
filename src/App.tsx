import axios from 'axios';
import { useEffect, useState } from 'react';
import LoadingSpinner from './LoadingSpinner';
import ErrorMessage from './ErrorMessage';
import RepositoryItem from './RepositoryItem';
import './App.css';

function App() {
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState("");
  const [topRepositories, setTopRepositories] = useState([]);

  useEffect(() => {
    setIsLoading(true);

    axios
      .get("https://api.github.com/search/repositories?q=stars:%3E1&sort=stars")
      .then((response) => {
        console.log(response.data.items);
        setTopRepositories(response.data.items);
      })
      .catch((err) => {
        setError(err.message);
      })
      .finally(() => {
        setIsLoading(false);
      });
  }, []);

  if (isLoading) {
    return <LoadingSpinner />;
  }

  if (error) {
    return <ErrorMessage message={error} />;
  }

  return (
    <div className="App">
      <h1>Top GitHub repos:</h1>
      <div className="repo-list">
        {topRepositories.map((repo: any) => (
          <RepositoryItem
            key={repo.id}
            name={repo.name}
            watchCount={repo.watchers}
            repositoryUrl={repo.html_url}
            homepageUrl={repo.homepage}
          />
        ))}
      </div>
    </div>
  );
}

export default App;
