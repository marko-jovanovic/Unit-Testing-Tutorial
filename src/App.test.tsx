import { create, act as rtrAct } from 'react-test-renderer'
import { render, screen, act, cleanup } from '@testing-library/react';
import axios from 'axios';
import App from './App';
import LoadingSpinner from './LoadingSpinner';
import ErrorMessage from './ErrorMessage';
import RepositoryItem from './RepositoryItem';

jest.mock('axios');

// Because Jest does monkey patatching behind the scene, this is something a TypeScript
// is not aware of, so, we need a way to "inform" TypeScript about this
const mockAxios = axios as jest.Mocked<typeof axios>;

describe('<App />', () => {
  beforeEach(() => {
    mockAxios.get.mockResolvedValue({
      data: {
        items: []
      }
    });
  });

  afterEach(cleanup);

  it('match snapshot', async () => {
    const component = create(<App />);

    // Since the mocked response is Promise-based we need to wait for an event loop to finish
    await act(async () => {
      await new Promise(resolve => setTimeout(resolve, 0));
    });

    expect(component.toJSON()).toMatchSnapshot();
  });
  
  it('renders learn react link', async () => {
    render(<App />);

    // Since the mocked response is Promise-based we need to wait for an event loop to finish
    await act(async () => {
      await new Promise(resolve => setTimeout(resolve, 0));
    });

    expect(screen.getByText('Top GitHub repos:')).toBeInTheDocument();
  });

  it('Should render LoadingSpinner component if request is in progress', () => {
    const componentInstance = create(<App />).root;

    // There is no need to do checkup, if component is not found the test will fail
    componentInstance.findByType(LoadingSpinner);
  });

  it('Should render ErrorMessage component if request fails', async () => {
    mockAxios.get.mockRejectedValue({ message: 'A terrible error ocurred :(' });
    const componentInstance = create(<App />).root;

    await rtrAct(async() => {
      await new Promise(resolve => setTimeout(resolve, 0));
    });

    componentInstance.findByType(ErrorMessage);
  });

  it('Should render 2 RepositoryItem components if request is successful', async () => {
    mockAxios.get.mockResolvedValue({
      data: {
        items: [
          {
            id: 28457823,
            watchers: 323864,
            homepage: 'https://contribute.freecodecamp.org',
            name: 'freeCodeCamp'
          },
          {
            id: 177736533,
            watchers: 256883,
            homepage: 'https://996.icu',
            name: '996.ICU'
          }
        ]
      }
    });
    const componentInstance = create(<App />).root;

    await rtrAct(async() => {
      await new Promise(resolve => setTimeout(resolve, 0));
    });

    const repositoryItems = componentInstance.findAllByType(RepositoryItem);
    expect(repositoryItems).toHaveLength(2);
  });
});
