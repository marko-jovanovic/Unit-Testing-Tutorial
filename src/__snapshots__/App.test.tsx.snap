// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> match snapshot 1`] = `
<div
  className="App"
>
  <h1>
    Top GitHub repos:
  </h1>
  <div
    className="repo-list"
  />
</div>
`;
